core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzerFactory.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.analyzer;
+
+import com.google.common.collect.ImmutableSet;
+import io.trino.Session;
+import io.trino.execution.warnings.WarningCollector;
+import io.trino.metadata.Metadata;
+import io.trino.security.AccessControl;
+import io.trino.spi.security.GroupProvider;
+import io.trino.sql.parser.SqlParser;
+import io.trino.sql.planner.TypeProvider;
+
+import javax.inject.Inject;
+
+import static java.util.Objects.requireNonNull;
+
+public class StatementAnalyzerFactory
+{
+    private final Metadata metadata;
+    private final SqlParser sqlParser;
+    private final AccessControl accessControl;
+    private final GroupProvider groupProvider;
+
+    @Inject
+    public StatementAnalyzerFactory(
+            Metadata metadata,
+            SqlParser sqlParser,
+            AccessControl accessControl,
+            GroupProvider groupProvider)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.sqlParser = requireNonNull(sqlParser, "sqlParser is null");
+        this.accessControl = requireNonNull(accessControl, "accessControl is null");
+        this.groupProvider = requireNonNull(groupProvider, "groupProvider is null");
+    }
+
+    public StatementAnalyzerFactory withSpecializedAccessControl(AccessControl accessControl)
+    {
+        return new StatementAnalyzerFactory(metadata, sqlParser, accessControl, groupProvider);
+    }
+
+    public StatementAnalyzer createStatementAnalyzer(
+            Analysis analysis,
+            Session session,
+            WarningCollector warningCollector,
+            CorrelationSupport correlationSupport)
+    {
+        return new StatementAnalyzer(this, analysis, metadata, sqlParser, groupProvider, accessControl, session, warningCollector, correlationSupport);
+    }
+
+    public ExpressionAnalyzer createExpressionAnalyzer(
```

Comment:
It seems weird to have a `createExpressionAnalyzer` in `StatementAnalyzerFactory`. Maybe the class deserves a better name.

Reply:
I added a commit at the end that removes this method... it isn't really possible before we added planner context.